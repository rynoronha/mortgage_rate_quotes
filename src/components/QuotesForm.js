import React, { useState } from 'react'

function QuotesForm() {
    const [formData, setFormData] = useState({
        loanSize: "",
        creditScore: "",
        propertyType: "SingleFamily",
        occupancy: "Primary"
    });

    const updateFormData = (event) => {
        setFormData({...formData, [event.target.name]: event.target.value});
    }

    const { loanSize, creditScore } = formData;

    const handleSubmit = (e) => {
        e.preventDefault();
        console.log(formData)
    }

    const formatNumberToCurrency = (e) => {
        var target = e.target;
        var temp = e.target.value.replace(/,/g, '');
        var value = Number(temp).toLocaleString('en-US');

        if (temp === '') {
            target.value = '';
        } else {
            target.value = '$' + value;
        }
    }

    const formatCurrencyToNumber = (e) => {
        let target = e.target;
        let value = target.value.replace(/[,.]/g, '');

        target.value = value.replace(/\$/,'');
    }

    return (
        <form className="form" onSubmit={e => handleSubmit(e)}>
            <div className="form-categories-container">
                <div className="form-category">
                    <label htmlFor="loan-size">Loan Size</label>
                    <input
                        value={loanSize}
                        onChange={e => updateFormData(e)}
                        onFocus={e => formatCurrencyToNumber(e)}
                        onBlur={e => formatNumberToCurrency(e)}
                        placeholder=""
                        type="text"
                        name="loanSize"
                        id="loan-size"
                        required
                    />
                </div>
                <div className="form-category select-container">
                    <label htmlFor="property-type">Property Type</label>
                    <select name="propertyType" id="property-type" onChange={e => updateFormData(e)}>
                        <option value="SingleFamily">Single Family</option>
                        <option value="Condo">Condo</option>
                        <option value="Townhouse">Townhouse</option>
                        <option value="MultiFamily">Multi Family</option>
                    </select>
                </div>
                <div className="form-category">
                    <label htmlFor="credit-score">Credit Score</label>
                    <input
                        value={creditScore}
                        onChange={e => updateFormData(e)}
                        placeholder=""
                        type="number"
                        name="creditScore"
                        id="credit-score"
                        min="300" 
                        max="850"
                        step="1" 
                        required
                    />
                </div>
                <div className="form-category select-container">
                    <label htmlFor="occupancy">Occupancy</label>
                    <select name="occupancy" id="occupancy" onChange={e => updateFormData(e)}>
                        <option value="SingleFamily">Primary Residence</option>
                        <option value="Condo">Secondary Residence</option>
                        <option value="Townhouse">Investment</option>
                    </select>
                </div>
            </div>
            <button className="button" type="submit">Quote Rates</button>
        </form>
    )
}

export default QuotesForm
