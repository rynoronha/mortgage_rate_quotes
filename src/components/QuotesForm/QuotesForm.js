import React, { useState } from 'react';
import { useDispatch } from "react-redux";
import { getQuotes } from "../../actions/quotes"
import { showLoader } from "../../actions/loader"

function QuotesForm() {
    const [formData, setFormData] = useState({
        loanSize: "",
        creditScore: "",
        propertyType: "SingleFamily",
        occupancy: "Primary"
    });

    const dispatch = useDispatch();
    
    const updateFormData = (event) => {
        if (event.target.name === "loanSize") {
            const re = /^[0-9\b]+$/;
           
            if (event.target.value === '' || re.test(event.target.value)) {
                setFormData({...formData, [event.target.name]: event.target.value});
            }
        } else {
            setFormData({...formData, [event.target.name]: event.target.value});
        }
        
    }
    const { loanSize, creditScore } = formData;

    const handleSubmit = (e) => {
        showLoader(dispatch);
        e.preventDefault();
        getQuotes(formData, dispatch);
    }

    const formatNumberToCurrency = (e) => {
        let target = e.target;
        let temp = e.target.value.replace(/,/g, '');
        let value = Number(temp).toLocaleString('en-US');

        if (temp === '') {
            target.value = '';
        } else {
            target.value = '$' + value;
        }
    }

    const formatCurrencyToNumber = (e) => {
        let target = e.target;
        let value = target.value.replace(/[,.]/g, '');

        target.value = value.replace(/\$/,'');
    }

    return (
        <form className="form" onSubmit={e => handleSubmit(e)}>
            <div className="form-categories-container">
                <div className="form-category">
                    <label htmlFor="loan-size">Loan Size</label>
                    <input
                        value={loanSize}
                        onChange={e => updateFormData(e)}
                        onFocus={e => formatCurrencyToNumber(e)}
                        onBlur={e => formatNumberToCurrency(e)}
                        placeholder=""
                        type="text"
                        name="loanSize"
                        id="loan-size"
                        required
                    />
                </div>
                <div className="form-category select-container">
                    <label htmlFor="property-type">Property Type</label>
                    <select name="propertyType" id="property-type" onChange={e => updateFormData(e)}>
                        <option value="SingleFamily">Single Family</option>
                        <option value="Condo">Condo</option>
                        <option value="Townhouse">Townhouse</option>
                        <option value="MultiFamily">Multi Family</option>
                    </select>
                </div>
                <div className="form-category">
                    <label htmlFor="credit-score">Credit Score</label>
                    <input
                        value={creditScore}
                        onChange={e => updateFormData(e)}
                        placeholder=""
                        type="number"
                        name="creditScore"
                        id="credit-score"
                        min="300" 
                        max="850"
                        step="1" 
                        required
                    />
                </div>
                <div className="form-category select-container">
                    <label htmlFor="occupancy">Occupancy</label>
                    <select name="occupancy" id="occupancy" onChange={e => updateFormData(e)}>
                        <option value="Primary">Primary Residence</option>
                        <option value="Secondary">Secondary Residence</option>
                        <option value="Investment">Investment</option>
                    </select>
                </div>
            </div>
            <button className="button" type="submit">Quote Rates</button>
        </form>
    )
}

export default QuotesForm
