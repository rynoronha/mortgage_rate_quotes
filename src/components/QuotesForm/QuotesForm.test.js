import React from 'react';
import  QuotesForm  from './QuotesForm';
import { render,  fireEvent, cleanup } from '../../test-utils'

afterEach(cleanup);

test('renders the form inputs and button', () => {
    const { queryByTestId } = render(<QuotesForm />);

    expect(queryByTestId('loan-size')).toBeTruthy();
    expect(queryByTestId('property-type')).toBeTruthy();
    expect(queryByTestId('credit-score')).toBeTruthy();
    expect(queryByTestId('occupancy')).toBeTruthy();
    expect(queryByTestId('button')).toBeTruthy();
});

test('it matches snapshot', () => {
    const tree = render(<QuotesForm />);

    expect(tree).toMatchSnapshot();
})

test('updates input values on change', () => {
    const { queryByTestId } = render(<QuotesForm />);

    const loanSizeInput = queryByTestId('loan-size');
    const creditScoreInput = queryByTestId('credit-score');

    fireEvent.change(loanSizeInput, {target: {value: '500000'}});
    fireEvent.change(creditScoreInput, {target: {value: '700'}});
    
    expect(loanSizeInput.value).toBe('500000');
    expect(creditScoreInput.value).toBe('700');
});

test('updates select values on change', () => {
    const { queryByTestId } = render(<QuotesForm />);

    const propertyTypeSelect = queryByTestId('property-type');
    const occupancySelect = queryByTestId('occupancy');

    fireEvent.change(propertyTypeSelect, {target: {value: 'Condo'}});
    fireEvent.change(occupancySelect, {target: {value: 'Investment'}});
    
    expect(propertyTypeSelect.value).toBe('Condo');
    expect(occupancySelect.value).toBe('Investment');
});

test('does not update loan size input value on change if non numeric character is entered', () => {
    const { queryByTestId } = render(<QuotesForm />);

    const loanSizeInput = queryByTestId('loan-size');

    fireEvent.change(loanSizeInput, {target: {value: 'test'}});
    
    expect(loanSizeInput.value).toBe('');
});

test('triggers handleSubmit function when form is submitted', () => {
    const spy = jest.spyOn(console, 'log')
    const wrapper = render(<QuotesForm />);

    const loanSizeInput = wrapper.queryByTestId('loan-size');
    const creditScoreInput = wrapper.queryByTestId('credit-score');

    fireEvent.change(loanSizeInput, {target: {value: '500000'}});
    fireEvent.change(creditScoreInput, {target: {value: '700'}});
    fireEvent.click(wrapper.queryByTestId('button'));

    expect(spy).toHaveBeenCalledTimes(1);
});

