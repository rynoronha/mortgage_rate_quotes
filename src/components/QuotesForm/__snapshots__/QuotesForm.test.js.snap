// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="form"
        data-testid="form"
      >
        <div
          class="form-categories-container"
        >
          <div
            class="form-category"
          >
            <label
              for="loan-size"
            >
              Loan Size
            </label>
            <input
              data-testid="loan-size"
              id="loan-size"
              name="loanSize"
              placeholder=""
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-category select-container"
          >
            <label
              for="property-type"
            >
              Property Type
            </label>
            <select
              data-testid="property-type"
              id="property-type"
              name="propertyType"
            >
              <option
                value="SingleFamily"
              >
                Single Family
              </option>
              <option
                value="Condo"
              >
                Condo
              </option>
              <option
                value="Townhouse"
              >
                Townhouse
              </option>
              <option
                value="MultiFamily"
              >
                Multi Family
              </option>
            </select>
          </div>
          <div
            class="form-category"
          >
            <label
              for="credit-score"
            >
              Credit Score
            </label>
            <input
              data-testid="credit-score"
              id="credit-score"
              max="850"
              min="300"
              name="creditScore"
              placeholder=""
              required=""
              step="1"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-category select-container"
          >
            <label
              for="occupancy"
            >
              Occupancy
            </label>
            <select
              data-testid="occupancy"
              id="occupancy"
              name="occupancy"
            >
              <option
                value="Primary"
              >
                Primary Residence
              </option>
              <option
                value="Secondary"
              >
                Secondary Residence
              </option>
              <option
                value="Investment"
              >
                Investment
              </option>
            </select>
          </div>
        </div>
        <button
          class="button"
          data-testid="button"
          type="submit"
        >
          Quote Rates
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="form"
      data-testid="form"
    >
      <div
        class="form-categories-container"
      >
        <div
          class="form-category"
        >
          <label
            for="loan-size"
          >
            Loan Size
          </label>
          <input
            data-testid="loan-size"
            id="loan-size"
            name="loanSize"
            placeholder=""
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="form-category select-container"
        >
          <label
            for="property-type"
          >
            Property Type
          </label>
          <select
            data-testid="property-type"
            id="property-type"
            name="propertyType"
          >
            <option
              value="SingleFamily"
            >
              Single Family
            </option>
            <option
              value="Condo"
            >
              Condo
            </option>
            <option
              value="Townhouse"
            >
              Townhouse
            </option>
            <option
              value="MultiFamily"
            >
              Multi Family
            </option>
          </select>
        </div>
        <div
          class="form-category"
        >
          <label
            for="credit-score"
          >
            Credit Score
          </label>
          <input
            data-testid="credit-score"
            id="credit-score"
            max="850"
            min="300"
            name="creditScore"
            placeholder=""
            required=""
            step="1"
            type="number"
            value=""
          />
        </div>
        <div
          class="form-category select-container"
        >
          <label
            for="occupancy"
          >
            Occupancy
          </label>
          <select
            data-testid="occupancy"
            id="occupancy"
            name="occupancy"
          >
            <option
              value="Primary"
            >
              Primary Residence
            </option>
            <option
              value="Secondary"
            >
              Secondary Residence
            </option>
            <option
              value="Investment"
            >
              Investment
            </option>
          </select>
        </div>
      </div>
      <button
        class="button"
        data-testid="button"
        type="submit"
      >
        Quote Rates
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
